### DISCUSSION TITLE
Python posting to PHP

### DISCUSSION THREAD

**Discussion by xRv9Fk0oQ6Da** [fastapi/fastapi#5055] [Questions] [answered]
Hi, I'm more experienced in PHP and I been trying to learn Python to help strengthen the security of my Software(s). 

However, I keep running into issues with the responses successfully connecting to a foreign PHP Script to post data.

Can someone please contribute a clone/edit version of: 
https://github.com/xRv9Fk0oQ6Da/fastapi-0003

Personally, it seems like it's impossible. 
I've tried with CORS and everything; nothing works.

Maybe if an experienced Python Dev. can kindly contribute an example, it'd be greatly appreciated for myself and other intermediate Python to PHP Devs who may be experiencing the same issue.

Thanks!

**Comment by acidjunk** [comment]
I peaked at the repo: Looked at the code in main.py, but I'm not sure what the question is...

Can you explain what you want to do?

What data do you want post; to what remote URI?

What seems impossible?

What did you try with CORS (Cors is a browser thing, so it confuses me to see it in a backend context)?

**Reply by acidjunk** [reply]
If you provide some details about the remote PHP backend that you want to post something to; I can probably help.

What's up with all the requirements?

`PyQt5-Qt5==5.15.2` + `Falcon` + a shitload of other deps: you only need FastAPI for the code that's in main.py.

If you want to build a simpel POST request I would suggest starting with just that; so completely outside of a framework whatsoever. Just install `requests` of `httpx` and write a simple function that you execute from the CLI. When that is working move it to a FastAPI endpoint. Now you ask people that want to help to install 40 deps just to run something that doesn't use these deps.

**Reply by xRv9Fk0oQ6Da** [reply]
Heyy @acidjunk, 
I'm new to Python and I followed some tutorials; but I was having issues getting it working. So, after I figured it out; I just threw it on a GitHub. The one I use has more routes and stuff; I'm still not to sure how the `freeze` works to remove pre-existing projects. Meaning, it kinda just bulks them all together. 

However, so the issue is when I trigger a post request; I want to forward its input from the Python to my PHP API/Endpoints and pages like "`.php?`"; when I do I have conversion issues where if it converts it doesn't form right- it's like both pieces cant compatibly work together. Consistent `422` if "successful"?

Please demonstrate a FastAPI post which can trigger to a `file.php` located at the main website which retains the PHP API. 

It's like they're having conflicting CORS Connections; I can't seem to figure it out. I've even tried different forms of JSON Parsing; nothing works. 

A simple FastAPI post to an external `https://example.com/file.php` doesn't work; however if it's set as `https://example.com/file.php?one=1&two=2` but only via `$_GET`. 

Also, apologies for the delay I've been at the hospital.

**Reply by acidjunk** [reply]
Normally when I add a requirement (like FastApi) -> I add it by hand. When you install FastApi ; it will also install the deps it needs.
`pip freeze` is used to inspects what is installed.

Have a look at my deps for a project that has quite some endpoint and DB stuff:
https://github.com/acidjunk/fastapi-postgres-boilerplate/blob/master/requirements/base.txt

A POST request can have a zillion options:

it can post a JSON payload, it has a path, upload a file, use authentication, SOAP, XML etc. Do you have some specs for the endpoint that you want to POST to?
Or a `postman` session?

As I mentioned before I would start as easy as possible: to it without FastAPI first; create a script that does the POST and run it form your CLI:

https://requests.readthedocs.io/en/latest/user/quickstart/#make-a-request

When that works -> copy it to an actual API endpoint in your fastapi app.

**Reply by xRv9Fk0oQ6Da** [reply]
Honestly, everything is kinda messy and I'm trying to have them all connect to each-other so I can begin stripping away the messy code. Currently, I don't have any due to constantly altering them doing as described. 

I'll update here later when I can have a moment to get on and try to figure it out. Unavailable at the moment. 

When returning a PHP format to Python, which would you recommend? 

I've been trying to figure it out with the `params=data` and JSON too. 

Should I `echo`, `print` or `return` from the external `.php` file?

Only way I get the `$_GET` (the `POST` from Python) to obtain a value is if I `echo json_encode($obj);` which I think could be the conversion issue.

Thanks for your help, it means a lot. 
Apologies if you downloaded anything unnecessary.

**Reply by xRv9Fk0oQ6Da** [reply]
So, my FastAPI post API looks like this:
```
@app.post("/post")
def getBinData(post: postSchema=None):
    r = requests.post("http://myapp.herokuapp.com/test.php?", params=post)
    r_dict = r.json()
    return r_dict
```
The external `.php` script I'm trying to post to from my Python Fast API. 
```
<?php
  header('Access-Control-Allow-Origin: *'); //for allow any domain, insecure
  header('Access-Control-Allow-Headers: *'); //for allow any headers, insecure

  header('Access-Control-Allow-Origin: http://localhost'); //for allow any domain, insecure
  header('Access-Control-Allow-Headers: http://localhost'); //for allow any headers, insecure

  header('Access-Control-Allow-Origin: https://my-api.herokuapp.com/post'); //for allow any domain, insecure
  header('Access-Control-Allow-Headers: https://my-api.herokuapp.com/post'); //for allow any headers, insecure

  header('Access-Control-Allow-Origin: https://myapp.herokuapp.com'); //for allow any domain, insecure
  header('Access-Control-Allow-Headers: https://myapp.herokuapp.com'); //for allow any headers, insecure

  header('Access-Control-Allow-Methods: GET, POST'); //method allowed
  header('Content-type: application/json; charset=utf-8');

    error_reporting(0);

      // echo $_GET['post'];

      if($_POST['post']) {
        $j = array("post"=>$_POST['post']);
        echo json_encode($j);
      } else if($_GET['post']) {
        $j = array("post"=>$_GET['post']);
        echo json_encode($j);
      } else {
        $j = array("error"=>"error occured");
        echo json_encode($j);
      }
?>
```
I want to take this data, in json from my app's UI, to my Python FastAPI to check on the received input from the UI then do the request, rather it any form of CRUD. I then want the PHP API to solve the query then return a json result.

**Comment by acidjunk** [comment, marked as answer]
Not sure what the "?" in you URI does, so removed it

```python
response = requests.post("http://myapp.herokuapp.com/test.php", json={"key": "value"})
if response.status_code == 200:
    # Assuming a json response
    print(response.json())
else:
    print(f"Error occured, status_code={response.status_code} received content={response.content}") 
```

**Reply by acidjunk** [reply]
Try that first from a python console/venv that has requests installed
